Modified Window Handling to be able to work over all active Browsers.

Snooz82 opened the pull request:
There are a few "Get" Keywords which return information about all windows of one browser. With this change it is also possible to get the window information of all windows of all browser.

Updated documentation about the relation of browsers and their windows.
Giving an example how to handle multiple browsers and multiple windows.

Cleanup:
Moving all keywords that where window related and not browser related from "BrowserManagementKeywords" to "WindowKeywords"

Snooz82 committed:
Added Keywords "Get Browser Ids" and "Get Browser aliases" to get a lists of all browsers.

Snooz82 committed:
Added "Switch Window" as replacement for select window.
Added "browser" Option to all wondow Getters to get windows information of all browsers.

Snooz82 committed:
Added Option "browser" to "Switch Window" (only CURRENT or <id or alias> are possible). Switches Browser before.

Snooz82 committed:
Updated Doc-Strings.
Created Browser/Window Documentation
Moved Window related Keywords from BrowsermanagementKeywords to WindowsKeywords

Snooz82 committed:
Added example for Browser and window handling in Doc

Snooz82 committed:
Select Window deprecation warning to silent.

Snooz82 commented:
fixes Issue: https://github.com/robotframework/SeleniumLibrary/issues/1388

Snooz82 commented:
@aaltat 
i moved these Keywords from BrowserManagementKeywords to WindowKeywords:
If you want, i can undo this and move it to another Pull-Request.

def maximize_browser_window(self):
def get_window_size(self, inner=False):
def set_window_size(self, width, height,   inner=False):
def get_window_position(self):
def set_window_position(self, x, y):
def get_source(self):
def get_title(self):
def get_location(self):
def location_should_be(self, url,   message=None):
def location_should_contain(self,   expected, message=None):
def log_location(self):
def log_source(self, loglevel='INFO'):
def log_title(self):
def title_should_be(self, title,   message=None):
def go_back(self):
def go_to(self, url):
def reload_page(self):

Snooz82 committed:
more improvements in documentation on window related keywords.

Snooz82 committed:
small improvement to Get Browser Aliases documentation.

aaltat commented:
I looked your code on mobile but I need to take a closer look with real PC. 

But in mean time, I think moving keywords did break some of the unit tests. Could you take a look what did actually break and fix it?

Snooz82 committed:
Moved window related keywords back to BrowserManagementKeywords to oversee the relevant changes better.

Snooz82 committed:
Fixed ordering get_session_id

Snooz82 committed:
Fixed Unittest to 173 Keywords instead of 170.

Snooz82 committed:
Set encoding to UTF-8 to be able to use all characters in Docstring

aaltat reviewed (changes requested):
That bash enforcement is not needed. Please remove it.

Snooz82 committed:
Fixed Issues with aTest

Commit:
Added acceptance tests for edited Window- and Browser-Information handling.
Modified dynamic_content.html to be able to manipulate the Window title freely.

Snooz82 commented:
@aaltat i modified the dynamic_content.html to be able to set the window title with a free text. in local Tests it had no negative effects to other atests.

I think it is ready to be reviewed... ;-)

Snooz82 committed:
Get rid of the for loop in atest because of Robot Framework 3.0.4 in Python 2.7 atests...

aaltat commented on src/SeleniumLibrary/keywords/browsermanagement.py:
```diff
@@ -350,14 +350,45 @@ def switch_browser(self, index_or_alias):
                    % self.driver.session_id)
 
+    @keyword
+    def get_browser_ids(self):
```
I did not understand where this keyword would be useful? Let's discuss.

Snooz82 replied:
The idea of this PR was to resolve complexity of window and browser handling.
1st reason for `get browser ids` and aliases was that we have corresponding Keywords for windows. This is for completeness.

2nd reason: It can happen, that browsers are opened in a very high level Suite Setup. Or that multiple browsers are open. If you want to Switch to a specific browser that has a specific site, you had no possibility to go through all Browsers. You could just iterate over some numbers, ` in range 0-10` but if some browsers are allready closed you will get problems.

aaltat replied:
I see your point. Then there should be documentation for the users how the keyword can be used with other keywords. There could be also tests to cover the usage scenario, so that we do not accidentally break the feature. 

Same comment goes with the `get_browser_aliases` comment.

Snooz82 replied:
It is hard for me to make an understandable pice of robot code.

Would be something like nested for loops. over browsers and windows.
And this is only possible with keywords in Keywords in robot.
Which makes it hard do document in Table.

Usecase would be switch to a window by its sontent. 
Of course i could do something like this:
```
Logging Window Titles for Each Browser
    @{browser_ids}=    Get Browser Ids
    FOR    ${id}    IN    @{browser_ids}
        @{window_titles}=    Get Window Titles    browser=${id}
        Log    Browser ${id} has these windows: ${window_titles}
    END
```

I could do this into the doc...
But it would not work as atest, because old Robot Version does not support this loop.

Snooz82 replied:
another example of how `Get Browser Ids` could be usefull:

```
*** Keywords ***
Switch Window By Location
    [Arguments]    ${selected_location}
    @{IDs}=    Get Browser Ids
    FOR     ${id}    IN    @{IDs}
        @{locations}=    Get Locations    browser=${id}   
        Run Keyword If    '${selected_location}' in $locations
        ...    Switch Window    url:${selected_location}    browser=${id}
    END
```

aaltat commented on src/SeleniumLibrary/keywords/browsermanagement.py:
```diff
@@ -350,14 +350,45 @@ def switch_browser(self, index_or_alias):
+        return self.drivers.active_driver_ids
+
+    @keyword
+    def get_browser_aliases(self):
```
I can see this be useful for logging purposes. but does this keyword server some other purpose too?

Snooz82 replied:
Logging and iterating over open browsers. 

Same as 'get id' this is also for be able to get what we set.

If some other tests opened a browser you can access with this keyword.

aaltat commented on src/SeleniumLibrary/keywords/browsermanagement.py:
```diff
@@ -350,14 +350,45 @@ def switch_browser(self, index_or_alias):
         """
         return self.driver.session_id
-
+    
```
Nit pic, spaces in here are not allowed.

Snooz82 replied:
I will fix this

Snooz82 replied:
Done

Snooz82 replied:
There are atests that are covering all of these Keyword!

I will do an example.

aaltat commented on src/SeleniumLibrary/keywords/window.py:
```diff
@@ -20,7 +20,7 @@
 from SeleniumLibrary.base import keyword, LibraryComponent
 from SeleniumLibrary.locators import WindowManager
-from SeleniumLibrary.utils import plural_or_not
+from SeleniumLibrary.utils import (plural_or_not, is_string)
```
Nit pic: Parenthesis hare are not needed.

Snooz82 replied:
Fixed

aaltat reviewed (commented):
I did not read all the code, but I have some questions first.

Snooz82 committed:
Cleaned up all Lines that are only made out of whitespaces.

Snooz82 committed:
fixed not needed parenthesis

aaltat commented on atest/acceptance/multiple_browsers_multiple_windows.robot:
```diff
@@ -0,0 +1,147 @@
+Resource          resource.robot
+
+*** Variables ***
+@{BrowserA_exp_Titles}=    WindowA1    WindowA2    WindowA3
```
Nit pick: we prefer to write global variables with capital letters. If it does not hurt readability, could global variables be written in upper case.

Snooz82 replied:
changed them to capital. 
But, i let "BrowserA" at This case because it is also the alias and used multiple times. So it is more a Name that a variable?! 
like BrowserA_EXP_TITLES

aaltat commented on src/SeleniumLibrary/__init__.py:
```diff
@@ -223,6 +224,89 @@ class SeleniumLibrary(DynamicCore):
+
+    == Browser ==
+
+    In Selenium Library there are two ways to start a Browser and control it.
```
Hmmm, I am hesitant about this chapter. Would it be better to explain something like this: 

_Each time the Open Browser or Create WebDriver keywords are called, it will create a new browser. Also Internally it creates a Selenium WebDriver_ [link here selenium docs would be nice] _instance. The new browser is, usually, independent from other browsers opened by the in the test, example they do not share cookies or example profile_

I would not explain the what keywords are preferred and about the keyword history, I think it confuses the reader. The above example is quite minimal and requires some more text.

Snooz82 replied:
Did refactor this chapter.
More details on Webdriver and Selenium Stuff.

aaltat commented on src/SeleniumLibrary/__init__.py:
```diff
@@ -223,6 +224,89 @@ class SeleniumLibrary(DynamicCore):
+    When opening a browser a new Webdriver process is startet and connected to
+    Selenium Library. When using Selenium Grid this browser may be started on any
+    Grid node that has matching capabilities.
+    It is possible to start multiple independent browsers at the same test case.
```
I like this chapter, let's keep this in.

Snooz82 replied:
Did refactor this chapter.
More details on Webdriver and Selenium Stuff.

aaltat commented on src/SeleniumLibrary/__init__.py:
```diff
@@ -223,6 +224,89 @@ class SeleniumLibrary(DynamicCore):
+    The second one is the `Create Webdriver` keyword which was useful, when
+    browser options needed to be set. Since Selenium Library 4.0.x `Open Browser`
+    also supports option why `Create Webdriver` will be deprecated.
+    When opening a browser a new Webdriver process is startet and connected to
```
Would it be good idea to have table, that describer the terms
| WebDriver | Selenium Python class to control browser |
Window, browser, browser driver (example chrome driver) kind a needs their own explanations. 

Also when explaining this thing, would I am starting to think that `browser` is not a good word. Instead it could be cleaner to talk about WebDriver, window and browser driver.

Snooz82 replied:
i think then it is starting to confuse people again. 

we could explain a lot...
Webdriver is a REST-API and Selenium is the inplementation and a Webdriver Server it the thing at the Browser and the client is Selenium. etc...
Then we could explaint, that Grid nodes are servers and the hub is a Client. 
And nobody would understand whats about windows and Browsers.
i would say, if people want to understand Selenium they should read their docs.

i tried to explain that a "browser" has "windows". I didnt want to explain that a Webdriver instance has a connection to a browserdriver etc...

Snooz82 replied:
Did refactor this chapter.
More details on Webdriver and Selenium Stuff.

aaltat commented on src/SeleniumLibrary/keywords/webdrivertools.py:
```diff
@@ -294,6 +294,18 @@ def active_drivers(self):
+        return open_driver_ids
+
+    @property
+    def active_aliases(self):
```
Instead of wrapping like this, it would be good to change `_alias` as a public method. `_alias` -> `alias`

Snooz82 replied:
But _aliases is a field of the Class ConnectionCache, which is part of robot.utils...
Do you wand me to make a pull request to this? This would brake a lot of libraries.
Also SSH Library uses _aliases etc...

May be i completely misunderstood.

aaltat replied:
I did not know that, no need to change.

Snooz82 commented:
> That bash enforcement is not needed. Please remove it.
removed...

Snooz82 committed:
Updating Doc. implementing review comments.

aaltat commented:
There is conflict that must be resolved before this can be merged.

Snooz82 committed:
Prepare Solve Conflict

Snooz82 committed:
prepare for solve conflicts

Snooz82 committed:
Merge branch 'master' of https://github.com/robotframework/SeleniumLibrary into robotframework-master

Snooz82 committed:
Merge branch 'robotframework-master'

Snooz82 committed:
Added an exapmle Test to switch to w window of another Browser by its Location using the Get Browser Ids Keyword

aaltat committed:
Updated documentation

aaltat commented:
I made modifications to the documentation, so that it would be more easier to understand. I did went to a direction that hides most the Selenium and browser driver stuff, because it felt confusing. Now it talks about browser and windows (mostly.) Now I am bit scared that I did lost something when I did cut down the details, because for me it is clear but it makes me blind to some problems. Should we still change it someway to make it better?

aaltat commented on src/SeleniumLibrary/locators/windowmanager.py:
```diff
@@ -39,7 +39,45 @@ def __init__(self, ctx):
+        if is_string(browser) and browser == 'ALL':
+            handles = []
+            current_index = self.drivers.current_index
+            for index, driver in enumerate(self.drivers, 1):
```
Hmm, are you sure that this will work, because drivers is a [set](https://docs.python.org/3.7/library/stdtypes.html#set-types-set-frozenset)

Snooz82 replied:
But set is iterable..

class set([iterable])

Or what exactly are you afraid is not working?

aaltat replied:
My bad, drivers are not in set. Please ignore
